gapis/gfxapi/vulkan/dependency_graph.go

```diff
@@ -0,0 +1,1733 @@
+		createInfo := typedAtom.PCreateInfo.Read(ctx, a, s, nil)
+		module := createInfo.Stage.Module
+		addRead(&b, g, vulkanStateKey(module))
+		addWrite(&b, g, vulkanStateKey(typedAtom.PPipeline.Read(ctx, a, s, nil)))
+
+	case *VkCreateShaderModule:
+		addWrite(&b, g, vulkanStateKey(typedAtom.PShaderModule.Read(ctx, a, s, nil)))
+
+	case *RecreateShaderModule:
+		addWrite(&b, g, vulkanStateKey(typedAtom.PShaderModule.Read(ctx, a, s, nil)))
+
+	case *VkCmdCopyImage:
+		addRead(&b, g, vulkanStateKey(typedAtom.SrcImage))
+		addRead(&b, g, vulkanStateKey(typedAtom.DstImage))
+		srcBindings := getOverlappedBindingsForImage(typedAtom.SrcImage)
+		dstBindings := getOverlappedBindingsForImage(typedAtom.DstImage)
+		addRead(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).handle)
+		addModify(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).records)
+		recordCommand(typedAtom.CommandBuffer, func(b *AtomBehaviour) {
+			readMemoryBindingsData(b, srcBindings)
+			// Be conservative here. Without tracking all the memory ranges and
+			// calculating the memory according to the copy region, we cannot assume
+			// this command overwrites the data. So it is labelled as 'modify' to
+			// kept the previous writes
+			modifyMemoryBindingsData(b, dstBindings)
+		})
+
+	case *RecreateCmdCopyImage:
+		addRead(&b, g, vulkanStateKey(typedAtom.SrcImage))
+		addRead(&b, g, vulkanStateKey(typedAtom.DstImage))
+		srcBindings := getOverlappedBindingsForImage(typedAtom.SrcImage)
+		dstBindings := getOverlappedBindingsForImage(typedAtom.DstImage)
+		addRead(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).handle)
+		addModify(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).records)
+		recordCommand(typedAtom.CommandBuffer, func(b *AtomBehaviour) {
+			readMemoryBindingsData(b, srcBindings)
+			// Be conservative here. Without tracking all the memory ranges and
+			// calculating the memory according to the copy region, we cannot assume
+			// this command overwrites the data. So it is labelled as 'modify' to
+			// kept the previous writes
+			modifyMemoryBindingsData(b, dstBindings)
+		})
+
+	case *VkCmdCopyImageToBuffer:
+		addRead(&b, g, vulkanStateKey(typedAtom.SrcImage))
+		addRead(&b, g, vulkanStateKey(typedAtom.DstBuffer))
+		srcBindings := getOverlappedBindingsForImage(typedAtom.SrcImage)
+		dstBindings := getOverlappedBindingsForBuffer(typedAtom.DstBuffer)
+		addRead(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).handle)
+		addModify(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).records)
+		recordCommand(typedAtom.CommandBuffer, func(b *AtomBehaviour) {
+			readMemoryBindingsData(b, srcBindings)
+			// Be conservative here. Without tracking all the memory ranges and
+			// calculating the memory according to the copy region, we cannot assume
+			// this command overwrites the data. So it is labelled as 'modify' to
+			// kept the previous writes
+			modifyMemoryBindingsData(b, dstBindings)
+		})
```

Comment:
Refactored in a way that combine `recordCommand` and memory binding `read/modify/write`
```
recordTouchingMemoryBindingsData := func(currentBehaviour *AtomBehaviour,
		handle VkCommandBuffer,
		readBindings, modifyBindings, writeBindings []*vulkanDeviceMemoryBinding) {
...
read(readBindings)....
}
```